zstd_stream.go

```diff
@@ -267,3 +268,24 @@ func (r *reader) Read(p []byte) (int, error) {
 	}
 	return got, nil
 }
+
+// TryReadFull reads buffer just as ReadFull does
+// Here we expect that buffer may end and we do not return ErrUnexpectedEOF as ReadAtLeast does.
+// We return ErrShortRead instead to distinguish short reads and failures.
+// We cannot use ReadFull/ReadAtLeast because it masks Reader errors, such as network failures
+// and causes panic instead of error.
+func TryReadFull(r io.Reader, buf []byte) (n int, err error) {
+	for n < len(buf) && err == nil {
+		var nn int
+		nn, err = r.Read(buf[n:])
+		n += nn
+	}
+	if n == len(buf) && err == io.EOF {
+		err = nil // EOF at the end is somewhat expected
+	} else if err == io.EOF {
+		err = ErrShortRead
+	}
+	return
+}
+
+var ErrShortRead = errors.New("short read")
```

Comment:
nit: can we put this at the beginning or the file.
Also since this error will not be returned (the `if` check if it's a shortRead and shallow), can we make it private ? `ErrShortRead` -> `errShortRead`